micro-rdk-installer/src/main.rs

```diff
@@ -269,6 +298,121 @@ fn main() -> Result<(), Error> {
+
+            log::info!("parsing ptable");
+            let ptable = PartitionTable::try_from_bytes(ptable_raw.clone()).unwrap();
+            log::info!(
+                "hash: {:?} \n partitions: {:?}",
+                old_ptable_hash,
+                ptable.partitions()
+            );
+
+            log::info!("Retrieving new image");
+            let tmp_new = tempfile::Builder::new()
+                .prefix("micro-rdk-bin")
+                .tempdir()
+                .map_err(Error::FileError)?;
+            let app_path_new = match &args.flash_args.bootloader {
+                Some(path) => PathBuf::from(path),
+                None => {
+                    let rt = Runtime::new().map_err(Error::AsyncError)?;
+                    rt.block_on(download_micro_rdk_release(&tmp_new, args.version.clone()))?
+                }
+            };
+
+            log::info!("extracting new partition table.");
+            let app_file_new = OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(app_path_new.clone())
+                .map_err(Error::FileError)?;
+            let _num_read = app_file_new
+                .read_at(&mut ptable_raw, PARTITION_TABLE_ADDR.into())
+                .map_err(Error::FileError)?;
+            let new_ptable_hash = md5::compute(ptable_raw.clone());
+
+            // Compare partition tables
+            if old_ptable_hash != new_ptable_hash {
+                log::error!("Partition tables do not match!");
+                log::error!("Rebuild and flash micro-rdk from scratch");
+                return Err(Error::BinaryEditError(64));
+            }
+
+            log::info!("partition tables match!");
+
+            log::info!("parsing ptable");
+            let ptable = PartitionTable::try_from_bytes(ptable_raw)
+                .map_err(|_| Error::PartitionTableError)?;
+            let nvs_partition_info = ptable
+                .find("nvs")
+                .expect("failed to retrieve nvs partition table entry");
+            let nvs_offset = nvs_partition_info.offset();
+            let nvs_size = nvs_partition_info.size();
+            log::debug!("nvs offset: {}, nvs_size: {}", nvs_offset, nvs_size);
+
+            // get binary size
+            let file_len = app_file_new.metadata().map_err(Error::FileError)?.len();
+            if (nvs_offset as u64 + nvs_size as u64) >= file_len {
+                return Err(Error::BinaryEditError(file_len));
+            }
+
+            // get nvs data from the old binary
+            let dir = tempfile::tempdir().unwrap();
+            let nvs_path = dir.path().join("nvs.data");
+            log::info!("extracting nvs data from device");
+            // let mut flasher = connect(&args.connect_args, &config, false, false).unwrap();
+            flasher
+                .read_flash(nvs_offset, nvs_size, 0x1000, 64, nvs_path.clone())
+                .map_err(Error::EspFlashError)?;
```

Comment:
stuck here for some reason. keep getting a `(CorruptData(4096, 4039)`, where `4096` is the read block size, and `4039` is the size of the chunk that was read.

This step is for retrieving the actual NVS data and not just partition info.
Trying to read the full image at the start (before parsing the running partition table) fails with this error as well.

Reply:
Note: related? 
https://github.com/espressif/esptool/issues/823#issue-1531203064

Reply:
is the stub properly loaded? https://docs.espressif.com/projects/esptool/en/latest/esp32/esptool/flasher-stub.html